### DISCUSSION TITLE
Some questions related to previous examples of Nooku Framework and Server

### DISCUSSION THREAD

**Discussion by alexandreelise** [joomlatools/joomlatools-pages#748] [General]
Hi @johanjanssens,

It was pure gold from architecture point of view. What really interested me is the part you implemented a REST level 2 API and implemented the OPTIONS verb with a working example. I tried to do it myself with Joomla! without much success for the OPTIONS verb maybe because I didn't read the spec as you did and say in the video. 

Another thing is that you manage to make communicate two joomla website using nooku server I found it nice as it work and don't rely on other external extension or tool. Since nooku server is your own code. But the drawback I see it that it is too centralized on one authoritative server. 

What would be nice is to make it work decentralized like for example mastodon. I made an attempt of doing that with a joomla system plugin called plg_system_jxj  pronounced J "x" J   joomla talks to joomla . https://github.com/alexandreelise/plg_system_jxj  But that plugin has at limitations: 

- First it has dependencies to com_api from techjoomla to use rest api for joomla 3 support 
- It relies also too heavily on joomla 4 new api and webservices. 
- It relies on the fact that we use it with joomla exclusively. 

A more agnostic and decentralized approach would be nice. Does the latest version of pages solves all this?

Thanks for your time and patience to answer this.

**Comment by johanjanssens** [comment]
Thats a trip down memory lane for sure. During JandBeyond 2011 (if i remember correctly, how time flies) we demonstrated a decentralised concept of making 2 Joomla sites communicate with each other, bridging sessions between sites. 

The code for this has since been improved further is now part of Joomlatools Framework and Pages. Pages offers all this and a lot more. I wrote a bit about the webservice support in Pages compared to Joomla 4 here: https://github.com/joomlatools/joomlatools-pages/discussions/675

The challenge with Pages now is to expose this power in a very easy and low-code approach. In essence with Pages I'm building a DSL [(Domain specific language)](https://github.com/joomlatools/joomlatools-pages/discussions/675) in YAML. 

The steps I'm working on are:

1. DSL for template engine 
2. DSL for read-only webservices (both importing data from other services as exposing data as a webservice)
3. DSL for read-write webservices (both writing to other services as writing to custom down service)

- **1** has been completed with the 0.21 release
- **2** is 95% done just need to to improve collection filtering (https://github.com/joomlatools/joomlatools-pages/issues/698)
- **3** is 50% done, it works but requires more work to expose the schema in yaml, and make it just work.

So yes, you can do all this with Pages, you either choose to do this by installing Pages in Joomla, or setup pages standalone and build your own app/webservice/...

For 3, I showed some examples in the talk i did a year ago for Joomlashack: https://www.youtube.com/watch?v=V13HaNKntuc
